204 entity auditing

simoncasey opened the pull request:
Added mysql config (look in application.yml for settings). Currently drops/creates db on each launch. Example revision history lookup at /profile/history.

eddiejaoude commented on src/main/java/io/dashboardhub/pipelinedashboard/config/ApplicationConfig.java:
```diff
@@ -0,0 +1,15 @@
+@Configuration
+public class ApplicationConfig {
+
+//    @Profile("!test")
```
Is this meant to be commented out?

simoncasey replied:
good point. missed this when sending PR. I'll cancel the pr

eddiejaoude replied:
Don't cancel the PR! Just update :smile:

eddiejaoude commented on src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProjectDefinitions.java:
```diff
@@ -10,6 +10,8 @@
 import org.junit.Assert;
 import org.springframework.beans.factory.annotation.Autowired;
 
+import java.util.UUID;
```
Is this needed?

simoncasey replied:
This is a leftover from when I was using proper UUIDs. It'll be needed when we use them, but otherwise it can just be removed.

eddiejaoude replied:
I would remove for now, to keep it cleaner. We can add back into the PR that requires them

eddiejaoude commented on src/main/java/io/dashboardhub/pipelinedashboard/domain/Project.java:
```diff
@@ -8,17 +11,15 @@
-    @Id
-    @GeneratedValue(strategy = GenerationType.AUTO)
-    private Long id;
+public class Project extends BaseEntity {
```
`final` got removed, is there a reason?

simoncasey replied:
merge issue. Still not sure why we want to mark all the entities as final though ...

eddiejaoude replied:
It's similar to `private` vs `public` reasons, make it *locked down* and *open up* as required. 

DI is better than Inheritance

eddiejaoude commented on src/main/java/io/dashboardhub/pipelinedashboard/domain/Project.java:
```diff
@@ -28,8 +29,13 @@
     private User user;
+
+    public Project() {
+        this.isPrivate = false;
```
Is it better to do it on the constructor that on the class property?

simoncasey replied:
I think so. It's easier to read IMO

eddiejaoude replied:
Ok

eddiejaoude commented:
We need **feature** tests to cover the changes.

simoncasey commented:
regarding feature tests - there were no user-facing changes made so it's a bit difficult to do feature tests for it (not including the revision history as that is a just an example that will not remain). Also, I'd like to pair with you on the use of the cucumber-style testing and to review our testing strategy (when we use unit, integration, feature)

eddiejaoude commented:
Why close it? Can just add to the **PR** :confused:

eddiejaoude commented:
Cucumber is not just for UI tests, it is just the abstraction. The FeatureContext determines the context, here we use the Context to check the DB https://github.com/DashboardHub/PipelineDashboard/blob/prototype-v0.3/src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProjectDefinitions.java#L22

eddiejaoude commented:
I did see a html page added, but even if there wasn't, the feature something like this:

```cucumber
  Scenario: Updating user entity should be audited 
    Given I have a User
    When I update the name
    Then I see a new audit entry
```

simoncasey commented:
ok. I added a feature to test the auditing and amended as per the little comments above. Also, changed the default (and therefore test) profile back to using H2 and changed travis to drop the mysql requirement. 

when running locally, if you want to use mysql, run the app in the dev profile.

eddiejaoude commented on src/main/java/io/dashboardhub/pipelinedashboard/config/ApplicationConfig.java:
```diff
@@ -0,0 +1,14 @@
+package io.dashboardhub.pipelinedashboard.config;
```
Application config is very generic, would it not be better to rename it to **DateConfig**?

eddiejaoude commented on src/main/java/io/dashboardhub/pipelinedashboard/config/CurrentTimeDateTimeService.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.time.ZonedDateTime;
+
+public class CurrentTimeDateTimeService implements DateTimeService {
```
If this class is called `CurrentTimeDateTimeService`, why is it in the **config** directory?

eddiejaoude commented on src/main/java/io/dashboardhub/pipelinedashboard/config/AuditingDateTimeProvider.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.Calendar;
+import java.util.GregorianCalendar;
+
+public class AuditingDateTimeProvider implements DateTimeProvider {
```
This file ends in **Provider**, should this not go into a **Provider** directory?

eddiejaoude commented on src/main/java/io/dashboardhub/pipelinedashboard/config/DateTimeService.java:
```diff
@@ -0,0 +1,8 @@
+
+import java.time.ZonedDateTime;
+
+public interface DateTimeService {
```
**Service** in the `config` directory?

eddiejaoude commented on src/main/java/io/dashboardhub/pipelinedashboard/config/UsernameAuditorAware.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+public class UsernameAuditorAware implements AuditorAware<String> {
```
I am not sure about the naming of the end of the file. Would it not be better placed in an `Auditor` directory?

eddiejaoude commented on src/main/java/io/dashboardhub/pipelinedashboard/config/UsernameAuditorAware.java:
```diff
@@ -0,0 +1,19 @@
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+        if (authentication == null || !authentication.isAuthenticated()) {
+            return null;
```
A note for the future, use the **NullObject** Pattern here

simoncasey replied:
Hmm, might be difficult here since we have no control over what the api returns. That said, this is a copy/paste from the tutorials on the net

eddiejaoude replied:
No worries, this is low priority for now :smile: ... more food for thought

eddiejaoude commented on src/main/resources/templates/account/history.html:
```diff
@@ -0,0 +1,32 @@
+            <table>
+                <tr th:each="revision : ${revisions}">
+                    <td th:text="${revision.getRevisionDate()}"></td>
+                    <td th:text="${revision.getEntity().toString()}"></td>
```
Also need a Feature test for this page

eddiejaoude commented on src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProfileDefinitions.java:
```diff
@@ -1,7 +1,61 @@
+    @Autowired
+    private UserRepository userRepository;
+
+    @Then("^There is a new revision for my User$")
```
No capital required otherwise would read `Then There...` but should be `Then there...`

eddiejaoude commented on src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProfileDefinitions.java:
```diff
@@ -1,7 +1,61 @@
+    private UserRepository userRepository;
+
+    @Then("^There is a new revision for my User$")
+    public void There_is_a_new_audit_entry_for_my_user() throws Exception {
```
No capital for method name

eddiejaoude commented on src/test/resources/io/dashboardhub/pipelinedashboard/definitions/profile/profile.feature:
```diff
@@ -25,3 +25,19 @@ Feature: User can update their profile
+      And I fill in the field name with updatedtestname
+      And I fill in the field email with updatedtest@email.com
+      And Submit the form form-profile
+    Then There is a new revision for my User
```
:+1:

eddiejaoude commented on src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProfileDefinitions.java:
```diff
@@ -1,7 +1,61 @@
+        Assert.assertTrue(latestRevision.getRevisionNumber() > 1);
+    }
+
+    @And("^the latest revised value for (.*) is (.*)$")
```
If under `Then` section, should be `@Then`. `And` is to generic and could be used in any section.

eddiejaoude replied:
I have changed it

eddiejaoude commented on src/test/java/io/dashboardhub/pipelinedashboard/definitions/ProfileDefinitions.java:
```diff
@@ -1,7 +1,61 @@
+        checkRevisionFieldValue(fieldName, value, 0);
+    }
+
+    @And("^the previous revised value for (.*) is (.*)$")
```
If under `Then` section, should be `@Then`. `And` is to generic and could be used in any section.

eddiejaoude replied:
I have changed it

eddiejaoude commented:
@simoncasey if you are happy with my additions, I am happy to merge the Pull Request

simoncasey commented:
@eddiejaoude sure go ahead and merge :)

eddiejaoude commented:
Cool. Lets do it! Merging...
